### PR TITLE
Refactor the replica scheduler

### PR DESCRIPTION
**Pull request by PhanLe1010**
* Refactor the replica scheduler for readability
* Allow disk eviction to move replicas from old disk to new disk on the same node even if the Node Level Soft Anti-Affinity is disable

longhorn/longhorn#3076

### CODE REVIEW

**Review by innobead — APPROVED** [review]
LGTM

**Comment by innobead** [comment]
@c3y1huang Please help review because this possibly impacts auto replica rebalancing.

**Comment by c3y1huang** [comment]
@PhanLe1010 , we should run `test_replica_auto_balance` to see if this change has an impact on auto replica rebalancing.

File: scheduler/replica_scheduler.go [hunk 1]
```diff
@@ -143,90 +165,82 @@ func (rcs *ReplicaScheduler) getDiskCandidates(nodeInfo map[string]*longhorn.Nod
...
+
+	filterNodesWithLessThanTwoReplicas := func(nodes map[string]*longhorn.Node) map[string]*longhorn.Node {
+		result := map[string]*longhorn.Node{}
+		for nodeName, node := range nodes {
+			if replicasCountPerNode[nodeName] < 2 {
+				result[nodeName] = node
 			}
 		}
+		return result
 	}
```
Comment (jenting) [inline scheduler/replica_scheduler.go:181]: Not quite understand this part's purpose.
Let's have a review meeting if possible.

**Review by jenting — COMMENTED** [review]
It will be good to make the function `getDiskCandidates` test coverage 100%

File: scheduler/replica_scheduler.go (hunk 1 above)
Comment (innobead) [reply scheduler/replica_scheduler.go:181]: This is the major closure to help #3076 when node soft anti-affinity is disabled to make sure being able to find out available extra disks to schedule on the same node where replicas all are being evicted. 

```go
		if diskCandidates := getDiskCandidatesFromNodes(filterNodesWithLessThanTwoReplicas(nodesWithEvictingReplicas)); len(diskCandidates) > 0 {
			return diskCandidates
		}
```

Comment (jenting) [reply scheduler/replica_scheduler.go:181]: I see. Then I am still thinking these two cases https://github.com/longhorn/longhorn-manager/blob/f9ee08e2e391dbba9a8d5fe096d6d19c99c39588/scheduler/replica_scheduler.go#L222-L241

**Review by jenting — APPROVED** [review]

File: scheduler/replica_scheduler.go (hunk 1 above)
Comment (PhanLe1010) [reply scheduler/replica_scheduler.go:181]: Discussed this question with @jenting over Slack.

Comment (PhanLe1010) [reply scheduler/replica_scheduler.go:181]: Created an issue for unit test implementation https://github.com/longhorn/longhorn/issues/3253

**Comment by innobead** [comment]
@PhanLe1010 please backport to 1.2.x.

**Pull request merged**